File: javascript/easy/validPerfectSquare.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function(num) {
    return search(1, num, num);
};

var search = function(start, end, num) {
    if(start > end) return false;
    let middle = parseInt((start + end)/2);
    let square = middle * middle;
    if(square === num) {
        return true;
    } else if(square > num) {
        return search(start, middle-1, num);
    } else {
        return search(middle+1, end, num);
    }
};

File: ruby/easy/valid_perfect_square.rb
# @param {Integer} num
# @return {Boolean}
def is_perfect_square(num)
    start, stop = 1, num
    while start <= stop
        middle = (start + stop)/2
        square = middle * middle
        if(square == num)
            return true
        elsif(square > num)
            stop = middle-1
        else
            start = middle+1
        end
    end
    return false
end
